src/util/file_watch.h

```diff
@@ -0,0 +1,41 @@
+/*
+ * Copyright (C) 2022 Red Hat Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#ifndef _FILE_WATCH_H
+#define _FILE_WATCH_H
+
+#include <stdbool.h>
+#include <talloc.h>
+#include <tevent.h>
+#ifdef HAVE_SYS_INOTIFY_H
+#include <sys/inotify.h>
+#endif
+
+
+typedef void (*fw_fn)(const char *filename, void *arg);
+struct file_watch_ctx;
+
+struct file_watch_ctx *fw_watch_file(TALLOC_CTX *mem_ctx,
```

Comment:
What is the reason to return `struct file_watch_ctx *`?

Currently user can't do anything with it. Imo, a `bool` - success/fail - would make more sense.

Perhaps something like `fw_cancel_watch()` would make some sense. In this case a handle would be needed.

Reply:
From the Design Page:

> No provision is done at this moment to provide a way to stop watching a watched file as it didn’t exist before. File watching is automatically stopped when the memory context is freed, which means the caller must not free this structure while monitoring is needed.

Returning a `bool` as you propose is possible, but requires that this pointer is kept somewhere so that it doesn't get lost. To achieve this, I just returned it to the caller and let it store the pointer in its context, which is kept for the process lifetime, but will be freed at the end (providing a clean shutdown)

In addition, returning the handle to the caller, enable us to one day create a `fw_cancel_watch()` function if it is ever needed.

Reply:
> but requires that this pointer is kept somewhere

On `TALLOC_CTX *mem_ctx` provided as a 1st arg?